presto-main/src/main/java/com/facebook/presto/server/security/CustomPrestoAuthenticator.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server.security;
+
+import com.facebook.airlift.http.server.AuthenticationException;
+import com.facebook.airlift.http.server.Authenticator;
+import com.facebook.presto.spi.security.AccessDeniedException;
+
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+
+import java.security.Principal;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+public class CustomPrestoAuthenticator
+        implements Authenticator
+{
+    private PrestoAuthenticatorManager authenticatorManager;
+
+    @Inject
+    public CustomPrestoAuthenticator(PrestoAuthenticatorManager authenticatorManager)
+    {
+        this.authenticatorManager = requireNonNull(authenticatorManager, "authenticatorManager is null");
+    }
+
+    @Override
+    public Principal authenticate(HttpServletRequest request)
+            throws AuthenticationException
+    {
+        try {
+            // Extracting headers into a Map
+            Map<String, List<String>> headers = getHeadersMap(request);
+
+            // Passing the header map to the authenticator (instead of HttpServletRequest)
+            return authenticatorManager.getAuthenticator().createAuthenticatedPrincipal(headers);
+        }
+        catch (AccessDeniedException e) {
+            throw new AuthenticationException(e.getMessage());
+        }
+        catch (RuntimeException e) {
+            throw new RuntimeException("Authentication failed due to an unexpected runtime error", e);
+        }
+    }
+
+    // Utility method to extract headers from HttpServletRequest
+    private Map<String, List<String>> getHeadersMap(HttpServletRequest request)
+    {
+        Map<String, List<String>> headers = new HashMap<>();
+        Enumeration<String> headerNames = request.getHeaderNames();
+        while (headerNames.hasMoreElements()) {
+            String headerName = headerNames.nextElement();
+            Enumeration<String> headerValues = request.getHeaders(headerName);
+            List<String> valuesList = new ArrayList<>();
+            while (headerValues.hasMoreElements()) {
+                valuesList.add(headerValues.nextElement());
+            }
+            headers.put(headerName, valuesList);
+        }
+        return headers;
```

Comment:
We can refactor this method and use Java Stream API

```suggestion
        return Collections.list(request.getHeaderNames())
            .stream()
            .collect(Collectors.toMap(
                    headerName -> headerName,
                    headerName -> Collections.list(request.getHeaders(headerName))
            ));
```